include/mujincontrollerclient/mujincontrollerclient.h

```diff
@@ -499,11 +496,33 @@ class MUJINCLIENT_API ControllerClient
 
     /// \brief Restore backup archive into controller. Restaring might be required after restoration.
     ///
-    /// \param inputStream the stream represententing the backup. It needs to be seekable to get the size (ifstream subclass is applicable to files).
+    /// \param inputStream the stream represententing the backup. It needs to be seekable to get the size (ifstream subclass is applicable for files).
     /// \param config whether to restore config. By default true (if the backup file has config).
     /// \param media whether to restore media. By default true (if the backup file has media).
     virtual void RestoreBackup(std::istream& inputStream, bool config = true, bool media = true, double timeout = 60.0) = 0;
 
+    /// \brief Upgrade controller's software.
+    ///
+    /// \param inputStream the stream represententing the software file provided by MUJIN.
+    ///        It needs to be seekable to get the size (ifstream subclass is applicable for files).
+    ///        NULL means using previously uploaded file.
+    /// \param autorestart whether to restart automatically after upgrading. if false, Reboot() can be called later.
+    /// \param uploadonly whether to upload only (so that upgrade can be continued later without uploading).
+    virtual void Upgrade(std::istream* inputStream, bool autorestart, bool uploadonly, double timeout = 600.0) = 0;
```

Comment:
@ziyan @kanbouchou @cielavenir

why not just pass in an empty std::istream?
And if its content is empty, then just don't add any form upload, which will just re-use the file.
Or is there a use case where we upgrade from empty file?

Either way, I don't think raw pointer is good here.

Reply:
I tested that (in current client) confirmed that uploading empty stream just causes error on hypervisor. Now I'll change the behavior on empty stream and remove pointer.

Reply:
New change looks good to me too. Is it easy to construct an empty istream?

Reply:
I see, `stringstream` will do.